### ISSUE TITLE
[Bug] Even after SparkContext stopped in yarn-client mode, Kyuubi plugin is fetching updated policies from Ranger. Spark driver does not end.

### ISSUE THREAD

**Issue by amanraj2520** [apache/kyuubi#6212]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

I am submitting a SparkSql job through spark-submit. Spark version 3.3.1 and Kyuubi version 1.8.0. I am using Open Source Spark Engine with Kyuubi Authz module running on top of the Spark Driver in client mode. The Spark job is successful, but the Spark Driver does not stop and keeps on running and I see the PolicyRefresher keeps on polling policies from Ranger.

![image](https://github.com/apache/kyuubi/assets/104416558/5bf7e9af-24d3-4ffb-8239-53eae0bd88fc)

If you see the logs, PolicyRefresher is still running even after Spark Context has stopped. This is leading to the Spark Driver not ending and therefore after sometime I have to manually kill the job.

Can someone please help in this issue?

### Affects Version(s)

1.8.0

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to fix.
- [ ] No. I cannot submit a PR at this time.

**Comment by pan3793** [comment]
Sounds like non-daemon threads blocks the Driver JVM to exit, you can find them from the jstack, I roughly remember that Ranger's solr and es audit plugin has such issues, not sure if they get fixed.

**Comment by amanraj2520** [comment]
Thanks @pan3793 for replying

"org.apache.ranger.audit.queue.AuditBatchQueue1-EventThread" #200 daemon prio=5 os_prio=0 tid=0x00007f1bec01b000 nid=0x4ca3 waiting on condition [0x00007f1bc3c03000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000007b20c9c88> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
        at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
        at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:559)

"org.apache.ranger.audit.queue.AuditBatchQueue1-SendThread(zk1-abc3.aadds.xyz.com:2181)" #199 daemon prio=5 os_prio=0 tid=0x00007f1bec019800 nid=0x4ca2 runnable [0x00007f1bc3d04000]
   java.lang.Thread.State: RUNNABLE
        at sun.nio.ch.EPollArrayWrapper.epollWait(Native Method)
        at sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
        at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:93)
        at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
        - locked <0x00000007b20c8050> (a sun.nio.ch.Util$3)
        - locked <0x00000007b20c8040> (a java.util.Collections$UnmodifiableSet)
        - locked <0x00000007b20c7c50> (a sun.nio.ch.EPollSelectorImpl)
        at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
        at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:332)
        at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1290)

"MdsLoggerSenderThread" #157 daemon prio=5 os_prio=0 tid=0x00007f1bd265c000 nid=0x4c52 sleeping[0x00007f1bc7027000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
        at java.lang.Thread.sleep(Native Method)
        at com.microsoft.mdsdclient.MessageSendingRunnable.run(Unknown Source)
        at java.lang.Thread.run(Thread.java:750)

"policyDownloadTimer" #138 daemon prio=5 os_prio=0 tid=0x00007f1d4feab800 nid=0x4c3d in Object.wait() [0x00007f1bf59e2000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        at java.util.TimerThread.mainLoop(Timer.java:552)
        - locked <0x00000006c2b54748> (a java.util.TaskQueue)
        at java.util.TimerThread.run(Timer.java:505)

"PolicyRefresher(serviceName=abc3-azsuu004-hive)-137" #137 daemon prio=5 os_prio=0 tid=0x00007f1d4feab000 nid=0x4c3c waiting on condition [0x00007f1c4416e000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000006c2b54838> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
        at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
        at org.apache.ranger.plugin.util.PolicyRefresher.run(PolicyRefresher.java:213)

I think you are right in the sense I see AuditBatch threads in the jstack. Anyway to fix this?

**Comment by pan3793** [comment]
@amanraj2520 all the above you listed is `daemon` threads, but the matter is those non-daemon threads

**Comment by amanraj2520** [comment]
"zkConnectionManagerCallback-5-thread-1" #202 prio=5 os_prio=0 tid=0x00007f1cbc003000 nid=0x4ca7 waiting on condition [0x00007f1bc3a01000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00000007b218dad8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
        at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
        at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)
        

"DestroyJavaVM" #453 prio=5 os_prio=0 tid=0x00007f1dfc019000 nid=0x4ab5 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   

"VM Thread" os_prio=0 tid=0x00007f1dfc09a000 nid=0x4ac3 runnable

"VM Periodic Task Thread" os_prio=0 tid=0x00007f1dfc10e800 nid=0x4ad4 waiting on condition

@pan3793 These are the non-daemon threads running. Any clues?

**Comment by pan3793** [comment]
`zkConnectionManagerCallback-5-thread-1` is likely created by Solr, that what I said the Ranger Solr audit plugin issue.

You may want to temporary disable Ranger Solr audit plugin to verify that.

**Comment by amanraj2520** [comment]
Thanks a ton for suggesting this @pan3793 . I will check this and get back to you in sometime. Really appreciate your prompt and useful response on this. I will update this thread accordingly and if needed will raise a PR to fix this as well.

**Comment by amanraj2520** [comment]
@pan3793 Thanks a lot for your suggestion. The issue of Spark Driver getting stuck is fixed now!!
The temporary fix was to set xasecure.audit.destination.solr to **false** thereby disabling logging to solr and in the jstack there was no more non-daemon thread running.

Having said this, it is our use case to enable Solr logging, is there any way that comes to your mind to fix solr side of things? Is there a discussion going on somewhere that you are aware of?

**Comment by pan3793** [comment]
This should be a Ranger/Solr issue and I recommend you report it to the Ranger community.

**Comment by amanraj2520** [comment]
Sure we can close this for now. Thanks again!!

**Comment by pan3793** [comment]
@amanraj2520 NP, and you are welcome to add cross-links here if you have better solutions to this problem.

**Comment by yaooqinn** [comment]
Perhaps we can enhance our documentation to assist others in understanding this problem and resolving it as well.

**Comment by amanraj2520** [comment]
Sure @yaooqinn Will do that.

@pan3793 After some debugging found that the SolrZkClient creates threads with daemon as false, that's why they are non-daemon threads 
![image](https://github.com/apache/kyuubi/assets/104416558/a0c14b53-4725-46c7-b762-658a838606c8)

So as per my understanding solr client runs on top of spark driver and when spark context is shutdown, ideally there should not be any solr client thread running right? Do you think making it as daemon as done for the PolicyRefresher thread in Ranger in this - https://issues.apache.org/jira/browse/RANGER-1083 work? I will give it a try and update this thread.

So in a nutshell the issue is this that the non-daemon thread of SolrZkClient are not exiting thereby leading to Spark Driver to run even if spark context is shutdown. Ideally this issue will not be for long running spark-shells or livy/zeppelin sessions, since they end by manual intervention of the end user. But in cases of spark-submit where the sparkContext has to run for a definite amount of time, SparkContext needs to have a way to end the SolrZkClient threads. This issue might not have impacted other services of Ranger (like hive, yarn etc) since they are long running sessions as well.

**Comment by pan3793** [comment]
@amanraj2520 Basically right, all third party components should use non-daemon threads in the Spark Driver. AFAIK, Spark YARN/K8s client mode and K8s cluster mode will hit non-daemon threads issues, but YARN cluster mode will not.

**Comment by amanraj2520** [comment]
@pan3793 So you think making SolrZkClient as daemon will not work?

**Comment by pan3793** [comment]
I think it will work

**Comment by amanraj2520** [comment]
Sure then will try my scenario with this patch

**Comment by amanraj2520** [comment]
@pan3793 Good news it worked by making solrZkClient as daemon thread. Thanks a lot for your help in this. I have filed for an OSS PR in SOLR as well - https://github.com/apache/solr/pull/2376

Raised a JIRA ticket for the same as well - https://issues.apache.org/jira/browse/SOLR-17220
